src/cli.rs

```diff
@@ -0,0 +1,821 @@
+  let _ = pretty_print(formatted_secret, &args.print_format);
+
+  Ok(())
+}
+
+async fn edit(client: impl QuerySecrets, args: &Edit) -> Result<(), NysmError> {
+  let secret_value = client.secret_value(args.secret_id.clone()).await?;
+  let secret = secret_value.secret;
+
+  if let Ok(dir) = temporary_directory() {
+    let update_contents = launch_editor(secret, dir, &args.secret_format, &args.edit_format)?;
+
+    if let Some(contents) = update_contents {
+      let _ = client
+        .update_secret_value(args.secret_id.clone(), contents)
+        .await?;
+    }
+  }
+
+  Ok(())
+}
+
+fn reformat_data(
+  content: &str,
+  source_format: &DataFormat,
+  destination_format: &DataFormat,
+) -> Result<String, NysmError> {
+  Ok(match source_format {
+    DataFormat::Json => {
+      let json_value: serde_json::Value = serde_json::from_str(content)?;
+
+      match destination_format {
+        DataFormat::Json => serde_json::to_string_pretty(&json_value)?,
+        DataFormat::Yaml => serde_yaml::to_string(&json_value)?,
+        DataFormat::Text => String::from(content),
+      }
+    }
+    DataFormat::Yaml => {
+      let yaml_value: serde_yaml::Value = serde_yaml::from_str(content)?;
+
+      match destination_format {
+        DataFormat::Json => serde_json::to_string_pretty(&yaml_value)?,
+        DataFormat::Yaml => serde_yaml::to_string(&yaml_value)?,
+        DataFormat::Text => String::from(content),
+      }
+    }
+    DataFormat::Text => String::from(content),
+  })
+}
+
+#[cfg(not(tarpaulin_include))]
+fn pretty_print(content: String, print_format: &DataFormat) -> Result<(), NysmError> {
+  if std::io::stdout().is_terminal() {
+    let language_string = print_format.to_string();
+    let mut printer = PrettyPrinter::new();
+    let _printer = match print_format {
+      DataFormat::Yaml | DataFormat::Json => printer.language(&language_string),
+      _ => &mut printer,
+    };
+
+    #[allow(unused)]
+    #[cfg(not(test))]
+    let _ = _printer
+      .grid(true)
+      .line_numbers(true)
+      .paging_mode(bat::PagingMode::QuitIfOneScreen)
+      .pager("less")
+      .theme("OneHalfDark")
+      .input_from_bytes(content.as_bytes())
+      .print()?;
+  } else {
+    println!("{}", content);
+  }
+
+  Ok(())
+}
+
+fn launch_editor(
+  contents: String,
+  path: tempdir::TempDir,
+  secret_format: &DataFormat,
+  edit_format: &DataFormat,
+) -> Result<Option<String>, NysmError> {
+  let language_string = edit_format.to_string().to_lowercase();
+  let file_path = path.path().join("data").with_extension(language_string);
+
+  let file_contents = reformat_data(&contents, secret_format, edit_format)?;
+  std::fs::write(&file_path, file_contents)?;
+
+  let editor = match std::env::var("EDITOR") {
+    Ok(editor) => editor,
+    Err(_) => String::from("vim"),
+  };
+  let editor_command = format!("{} {}", editor, &file_path.to_string_lossy());
+
+  #[cfg(test)]
+  let editor_command = format!("vim(){{ :; }}; {}", editor_command);
```

Comment:
Kind of syntactical semantics, but you could make `editor` mutable, then just push the file path string to it. And for the test code, you can prepend with:
```rust
editor_command.insert_str(0, "vim(){{ :; }}; ");
```

Reply:
I can see it working both ways. I think my preference is to leave it using string formatting. Given we are building out shell commands, I think it's easier to see what's happening visually. But I could be convinced to change it if there's some benefit I'm not seeing.

Reply:
I see only 1 actual reason for changing it. Each variable is going to be allocating memory for its own version of the command. However in this case, the strings involved are incredibly small. Meaning you can just use whichever you think expresses intent the best.

Reply:
Here's what it could look like using a mutable editor:

```rust
  let mut editor = match std::env::var("EDITOR") {
    Ok(editor) => editor,
    Err(_) => String::from("vim"),
  };

  editor.push(' ');
  editor.push_str(&file_path.to_string_lossy());

  #[cfg(test)]
  editor.insert_str(0, "vim(){ :; }; ");

  std::process::Command::new("/usr/bin/env")
    .arg("sh")
    .arg("-c")
    .arg(editor)
    .spawn()
    .expect("Error: Failed to run editor")
    .wait()
    .expect("Error: Editor returned a non-zero status");
```

Thoughts?

Reply:
This way feels cleaner to me. I quite like the idea that some base command is being built upon, rather than creating a new one for every step. I do feel like the line for tests should have some added context through a comment though. It's not entirely obvious what's going on there, and it feels a bit out of place. 
Something like: "Mask the vim command with a function to prevent vim from opening during tests."

Reply:
I don't really think a comment is necessary in this instance. If you know we are building out a shell command, you can see we are adding a shell function to intercept vim and do nothing. Even if you don't know the `:;` stuff inside the function definition, I think it's fairly obvious it's doing nothing.